org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/execution/UpdateExecutionPageJob.java

```diff
@@ -34,15 +33,18 @@
 
     private static final int REPEAT_DELAY = 100;
 
-    private final Map<OperationDescriptor, OperationItem> operationItems;
+    private final Set<OperationItem> operationItems;
+    private final Set<OperationItem> removeableItems;
+
     private volatile boolean running;
     private final TreeViewer treeViewer;
 
     public UpdateExecutionPageJob(TreeViewer treeViewer) {
         super("Updating duration of non-finished operations");
 
         this.treeViewer = treeViewer;
-        this.operationItems = Maps.newHashMap();
+        this.operationItems = Sets.newConcurrentHashSet();
```

Comment:
A general caution about concurrency: Using two concurrent sets doesn't guarantee safety when these two sets have a combined meaning/need to be coordinated. In this special case it is probably fine because we know each object is unique and the started and end events always come from the same thread in the right order. But I haven't thought too deeply about it, so there might still be a race condition I'm missing.

Unless the synchronization overhead is actually a measurable problem, I would suggest to go with a simple synchronized block around the whole access to both of these fields.